### PR TITLE
Killing the Discovery Service started by the Test Environment

### PR DESCRIPTION
**Pull request by MounikaBattu17**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
Acceptance tests (`test_measurement_service.py, test_streaming_data_measurement.py, test_yield_vs_return.py`) starts the discovery service for running their tests but doesn't terminate/kill it. This causes a `Resource Warning` while running all tests.
Approach is to start the Discovery service independently and scope it to whole test session.
1) Created a Context Manager  `DiscoveryServiceProcess` in `tests/utilities/discovery_service_process.py`.
      - `__init__` is responsible for starting the discovery service and store the instance in `self._proc`.
      - `__exit__` is responsible for killing the active Discovery service instance.
2) Added a pytest fixture `discovery_service_process` that is responsible for 
     - skipping tests for non-windows and measurement link registration file not found.
     -  yielding the `proc` and scoped it to `session` for windows platform.
3) Called the `discovery_service_process` fixture in `measurement_service` fixture in each acceptance test files.

### Why should this Pull Request be merged?
1) Starts and Closes Discovery Service instances.
2) Fixes Resource Warning while running tests.

### What testing has been done?
1) Ran `poetry run pytest`, Observed there is no `Resource Warning`.
2) Observed Discovery Service is started once and closed at the end of the test session.
3) Verified acceptance tests are skipped if platform is non windows or measurement link registration file not available.

### CODE REVIEW

**Review by vigkre — APPROVED** [review]

File: ni_measurementlink_service/_internal/utilities/globaltestingstate.py
Comment (bkeryan) [inline ni_measurementlink_service/_internal/utilities/globaltestingstate.py:1]: I think you should delete this file and use a different approach. 

However, I want to share my feedback on this file anyway:
- I do not like `_internal`. Going forward, I would prefer to prefix module names with `_` to indicate that they are internal.
- I think `utilities` is too general and does not make a meaningful distinction.
- The `GlobalTestingState` class does not need to exist. Python allows global variables at file/module scope.
- `IsInTestState` isn't following https://pep8.org/#method-names-and-instance-variables
- The wording of `IsInTestState` doesn't really make sense to me. The corresponding C# property is `IsInTestCode`, not `IsInTestCode`.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [inline tests/utilities/discovery_service_process.py:1]: This is a context manager, not a generator.

```suggestion
"""Class to create and terminate Discovery Service instance."""
```

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [inline tests/utilities/discovery_service_process.py:14]:
```suggestion
    def __init__(self) -> None:
```

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [inline tests/utilities/discovery_service_process.py:26]: Add one of these (preferably the 1st, if it works):
- `from typing_extensions import Self`
- `Self = TypeVar("Self", bound=object)`

```suggestion
    def __enter__(self: Self) -> Self:
```

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [inline tests/utilities/discovery_service_process.py:30]:
```suggestion
    def __exit__(self, exc_type: Optional[Type[BaseException]], exc_val: Optional[BaseException], exc_tb: Optional[TracebackType]) -> None:
```

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [inline tests/utilities/discovery_service_process.py:34]:
```suggestion
    def _close_discovery_service(self) -> None:
```

File: tests/acceptance/test_yield_vs_return.py
Comment (bkeryan) [inline tests/acceptance/test_yield_vs_return.py:58]: Did ni-python-styleguide warn about the unused parameter?

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: ni_measurementlink_service/_internal/utilities/globaltestingstate.py
Comment (bkeryan) [reply ni_measurementlink_service/_internal/utilities/globaltestingstate.py:1]: The new fixture looks good.

File: tests/acceptance/test_yield_vs_return.py
Comment (MounikaBattu17) [reply tests/acceptance/test_yield_vs_return.py:58]: I ran `poetry run ni-python-styleguide lint`, no errors.

Comment (mshafer-NI) [reply tests/acceptance/test_yield_vs_return.py:58]: a: assigining it to "_" (throwing it away) is referencing it, 
b: unused parens are allowed.

Remove this assignment.

**Review by vigkre — APPROVED** [review]

**Comment by dixonjoel** [comment]
Please update the PR description to match the implementation since you're not using the `IsInTestState` global now.

File: tests/utilities/discovery_service_process.py
Comment (MounikaBattu17) [inline tests/utilities/discovery_service_process.py:29]: @bkeryan The pipeline(Run Tests for all platforms, all acceptance tests) errors if I raise the exception here. There are 2 errors that occurs in `discovery_client.py`
1) Non windows platform - `NotImplementedError("Platform not supported")` error from `_get_key_file_directory()`
2) windows platform - FileNotFound error from `_get_discovery_service_location()`

`discovery_client.py` handles these exceptions by logging them.

Due to these errors, Discovery Service is not even started and the acceptance tests pass showing that these tests are independent of Discovery Service running or not

So, my question is 
1) Whether we should validate the loggers for appropriate error messages in acceptance tests? or
2) Since the tests are independent on Discovery Service, can we fake the discovery service stub?

File: tests/acceptance/test_yield_vs_return.py
Comment (bkeryan) [reply tests/acceptance/test_yield_vs_return.py]: @MounikaBattu17 Why are you still explicitly referencing the unused parameter?

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [reply tests/utilities/discovery_service_process.py:29]: Another option is to have the fixture call pytest.skip if the discovery service is not installed, like nidaqmx-python does. This is pretty easy to do and would unblock this PR, but we still need to figure out our strategy for testing Python code that uses real MeasurementLink core services.

File: tests/acceptance/test_yield_vs_return.py
Comment (MounikaBattu17) [reply tests/acceptance/test_yield_vs_return.py]: This file change is outdated. I have removed the unused params(`_`). Please refer to the new change.

File: tests/utilities/discovery_service_process.py
Comment (MounikaBattu17) [reply tests/utilities/discovery_service_process.py:29]: @bkeryan we want some clarifications on the Acceptance tests. So as we know acceptance tests starts and tries to register the measurements to the discovery service, 
If Discovery Service  is successfully started, it is not terminated
If Discovery Service  doesn't start successfully, exceptions are logged and tests pass

In the second case, the tests passing even though the measurement is not even registered to the discovery service is conveying that the overall workflow is not tested in the "Acceptance Tests" and I assume that Acceptance tests are end to end testing that tests the complete workflow of running a measurement. Please correct me If I'm wrong here.

So, we want to make the discovery service up and running for windows (which is only supported) such that acceptance tests works with the discovery service. To achieve this we are proposing the following approaches:

**1) Make Acceptance tests run only for windows:**
Mark all acceptance tests as windows only and start and terminate the discovery service using the `discovery_service_process` fixture. Doing this we will be testing the complete workflow needed to run a measurement.
(or)
**2) Use Fake discovery client stub for non windows platforms:**
Run acceptance tests for all platforms and use a fake discovery service if it is a non windows platform and use `discovery_service_process` for windows.
(or)
**3) Add new acceptance tests to verify discovery logs:**
Add new acceptance tests to verify the logs of Discovery service, thus we can test the error logs for non windows platform and success logs for windows platform

Please share your views on the approaches and direct on what approach to take or any other different direction will be helpful.

Ps: If we are going for approach 1, there is an `FileNotFound` error in pipeline, we are unsure of why that's occurring. please help us in debugging it.

**Comment by MounikaBattu17** [comment]
> @MounikaBattu17 Did #344 not fix the ResourceWarning?

Yes, It fixed the resource warning. Now the ResourceWarning doesn't come if I run a measurement.

**Comment by bkeryan** [comment]
> This causes a `Resource Warning` while running all tests.

If the resource warning was already fixed, why does the description of this PR talk about the resource warning?

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [reply tests/utilities/discovery_service_process.py:29]: > 1) Make Acceptance tests run only for windows:
Mark all acceptance tests as windows only and start and terminate the discovery service using the discovery_service_process fixture. Doing this we will be testing the complete workflow needed to run a measurement.

> Ps: If we are going for approach 1, there is an FileNotFound error in pipeline, we are unsure of why that's occurring. please help us in debugging it.

```
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Program Files\\National Instruments\\Shared\\MeasurementLink\\MeasurementLinkServices.json'
```

The GitHub-hosted runner does not have MeasurementLink installed, so the JSON registration file does not exist. Note that this is separate from the key file. The registration file is created by the MeasurementLink installer, and the key file is created by the discovery service itself.
 
For this PR, I think it is reasonable for the discovery service test fixture to call pytest.skip in these situations:
- sys.platform != "win32"
- JSON registration file does not exist

I think skipping in the test fixture is better than adding `@pytest.skipif(sys.platform != "win32", reason="...")` to a bunch of tests.

> 2) Use Fake discovery client stub for non windows platforms:
Run acceptance tests for all platforms and use a fake discovery service for a non windows platform and use discovery_service_process for windows.

This adds more non-Windows test coverage but it's more effort/complication than skipping the tests on non-Windows platforms.

> 3) Add new acceptance tests to verify discovery logs:
Add new acceptance tests to verify the logs of Discovery service, thus we can test the error logs for non windows platform and success logs for windows platform

I don't think this is worth it. I would rather that we improve the error handling so that it's not necessary to check the logs for errors.

> Please share your views on the approaches and direct on what approach to take or any other different direction will be helpful.

I like this approach best:
- Now: have discovery fixture skip on non-Windows
- Now: have discovery fixture skip when MeasurementLink is not installed (registration file not found)
- Separately: improve error handling to exit when registration fails
- Separately: add self-hosted GitHub runner w/MeasurementLink installed (note that we might have a simpler way to do this than what we did for nidaqmx-python)

@dixonjoel @pbirkhol-ni

**Review by Tharun-Sundar — APPROVED** [review]

File: tests/conftest.py
Comment (vigkre) [inline tests/conftest.py]: @MounikaBattu17  Can we use `pathlib.Path.exists` function here?

**Review by vigkre — APPROVED** [review]

File: tests/conftest.py
Comment (mshafer-NI) [inline tests/conftest.py:58]:
```suggestion
    elif not pathlib.Path(_get_registration_json_file_path()).exists():
```

This is if _get_registration_json_file_path() returns a str, if it returns a Path object, just do

```suggestion
    elif not _get_registration_json_file_path().exists():
```

File: tests/utilities/discovery_service_process.py
Comment (MounikaBattu17) [reply tests/utilities/discovery_service_process.py:29]:
```
Now: have discovery fixture skip on non-Windows
Now: have discovery fixture skip when MeasurementLink is not installed (registration file not found)
```
Thanks @bkeryan for the approach. I have added if conditions in `discovery_service_process` fixture in `conftest.py` to skip tests for  non-Windows and when MeasurementLink is not installed (registration file not found) with appropriate messages.

As per the other items:
`Separately: improve error handling to exit when registration fails` - This will be addressed in the upcoming PR.

`Separately: add self-hosted GitHub runner w/MeasurementLink installed (note that we might have a simpler way to do this than what we did for nidaqmx-python)` - I have created a **[Technical Debt 2486477](https://dev.azure.com/ni/DevCentral/_workitems/edit/2486477): Add Self hosted GitHub runner to run tests in measurementlink-python pipeline** to track.

File: tests/conftest.py
Comment (MounikaBattu17) [reply tests/conftest.py]: The return type is a `Path` object, So I have used the second suggested change.

**Comment by MounikaBattu17** [comment]
> > This causes a `Resource Warning` while running all tests.
> 
> If the resource warning was already fixed, why does the description of this PR talk about the resource warning?

The resource warning which I'm referring is from the tests. Earlier/ without this PR changes, there is a resource warning after the test run.

File: tests/utilities/discovery_service_process.py
Comment (dixonjoel) [reply tests/utilities/discovery_service_process.py:29]: We have 'spawningpool' that we use for running the LabVIEW tests: https://github.com/ni/measurementlink-labview/settings/actions/runners
It's already set up as a self-hosted runner. I wonder if it's feasible to use it on two repos.

Comment (dixonjoel) [reply tests/utilities/discovery_service_process.py:29]: In the meantime, it seems like we should be running the skipped tests locally with MeasurementLink installed before submitting PRs.

**Review by dixonjoel — APPROVED** [review]
All tests pass locally for me and don't get skipped when I have the DS installed.

File: tests/acceptance/test_yield_vs_return.py
Comment (bkeryan) [reply tests/acceptance/test_yield_vs_return.py]: Why is this still in the code?

```
@pytest.fixture(scope="module")
def measurement_service(discovery_service_process) -> Generator[MeasurementService, None, None]:
    """Test fixture that creates and hosts a measurement service."""
    discovery_service_process  # <- THIS
    with yield_vs_return_measurement.measurement_service.host_service() as service:
        yield service
```

File: tests/acceptance/test_measurement_service.py
Comment (bkeryan) [reply tests/acceptance/test_measurement_service.py:210]: See this discussion: https://github.com/ni/measurementlink-python/pull/346#discussion_r1289046069

File: tests/acceptance/test_yield_vs_return.py
Comment (bkeryan) [inline tests/acceptance/test_yield_vs_return.py]:
```suggestion
```

File: tests/acceptance/test_streaming_data_measurement.py
Comment (bkeryan) [inline tests/acceptance/test_streaming_data_measurement.py]:
```suggestion
```

File: tests/acceptance/test_measurement_service.py
Comment (bkeryan) [inline tests/acceptance/test_measurement_service.py]:
```suggestion
```

File: tests/utilities/discovery_service_process.py
Comment (bkeryan) [reply tests/utilities/discovery_service_process.py:29]: Skipping tests is sufficient for this PR.

Adding a self-hosted runner is a matter for another PR, and we probably need to create a work item for it too.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/acceptance/test_yield_vs_return.py
Comment (MounikaBattu17) [reply tests/acceptance/test_yield_vs_return.py]: Sorry for the confusion. I misinterpreted wrongly and removed only the left hand side assignment. Now I have corrected.

File: tests/acceptance/test_measurement_service.py
Comment (MounikaBattu17) [reply tests/acceptance/test_measurement_service.py]: Sorry for the confusion. I removed the unused call since it is dependency injected.

**Comment by MounikaBattu17** [comment]
This PR has got necessary approval. Hence merging.

**Pull request merged**
